### PR TITLE
rfq+rfqmsg: add fill quantity negotiation and constraint validation

### LINKED ISSUE #2004: [feature]: RFQ: First-Class Limit Orders In Quote Negotiation
## Summary
Today, RFQ orders are mostly constrained by:
- Requested max size (`asset_max_amt` or `payment_max_amt`)
- Quote expiry
- Channel availability and balance shape
- Price-oracle tolerance checks

This is sufficient for basic RFQ flows, but it is not a full limit-order model. We should add explicit limit-order semantics (price bounds, fill constraints, and execution policy) to quote negotiation.

## Why This Matters
Limit-order semantics enable:
- Price protection: avoid execution outside user-defined acceptable price bounds.
- Slippage control: reduce risk from stale hints, spread moves, and short-term volatility.
- Deterministic UX: users can express explicit execution intent instead of best-effort behavior.
- Safer automation: wallet agents can auto-negotiate with tighter risk controls.
- Better policy explainability: clearer machine-readable reject reasons.
- Better multi-peer execution: easier to split across peers while preserving user constraints.

## Terminology
- `IOC` (`Immediate-Or-Cancel`): execute immediately for any amount that satisfies constraints; cancel any unfilled remainder.
- `FOK` (`Fill-Or-Kill`): execute immediately and fully, or cancel the entire order.

### What RFQ can express today
- Buy flow: max asset amount (`AssetMaxAmt`) + expiry + peer + metadata.
- Sell flow: max payment amount (`PaymentMaxAmt`) + expiry + peer + metadata.
- Request/accept negotiation primarily settles on a rate + expiry.

### What is enforced in practice
- Requester-side quote validation checks include quote expiry bounds and oracle tolerance against the accepted rate.
- HTLC/policy enforcement is max-cap based, where sale path policy derives from `AssetMaxAmt` and purchase path policy derives from `PaymentMaxAmt`.
- Liquidity constraints are effectively channel-driven: channel existence checks gate order submission, and channel balances influence selection/sorting.

### Existing but underused protocol surface
- Wire format already includes optional `MinInAsset` and `MinOutAsset` fields.
- These min fields are not first-class in typed request structs and RPC order APIs.

### Current gap
Current behavior is closer to:
- "Market quote with max-size cap + expiry"
not:
- "Limit order with explicit price/fill constraints"

Users cannot directly encode:
- Minimum acceptable price / rate.
- Maximum acceptable price / rate.
- Minimum fill size.
- Execution policy (`IOC` vs `FOK`).

### Workstream A: Add explicit limit constraints to request surfaces
- Add optional limit-price/rate bound fields to buy/sell order APIs.
- Add optional fill bound fields (`min_in_asset`, `min_out_asset`) to typed requests and RPC.
- Keep current max fields as-is for compatibility.
- Define a canonical rate-bound representation that avoids buy/sell direction ambiguity.

### Workstream B: Enforce limits during negotiation
- Responder side (`ResolveRequest`): reject when request limits cannot be satisfied.
- Requester side (`VerifyAcceptQuote`): enforce explicit requester limits in addition to oracle tolerance.
- Return explicit reject reasons for limit failures (price bound miss, min fill miss, insufficient liquidity).

### Workstream C: Add execution policy semantics
- Support explicit execution policy on orders.
- `IOC`: allow partial immediate execution above minimum fill.
- `FOK`: require full immediate execution or reject.
- Ensure policy is respected consistently in quote acceptance and HTLC policy installation.

### Workstream D: Negotiate explicit fill quantities
- Add accepted fill quantity semantics to quote acceptance path.
- Bind policy persistence and HTLC enforcement to negotiated fill values, not only request max caps.
- This removes ambiguity around partial fills and improves deterministic execution.

### Workstream E: Multi-peer selection that respects limits
- When multiple peers are available, rank and split using effective rate and usable capacity.
- Ensure every shard respects the same order limits and execution policy.

## Non-Goals
- Building a centralized order book.
- Long-lived resting/GTC orders in initial implementation.
- Redesigning the external price-oracle architecture.

## Compatibility Expectations
- New fields should be additive and optional.
- Old peers should continue to interoperate under current behavior.
- Limit-specific validation should only activate when limit fields are present.
- Use versioning only where strict incompatibility requires it.

## Acceptance Criteria
- Users can specify explicit limit-price bounds for buy and sell RFQ orders.
- Users can specify minimum fill constraints.
- Users can choose execution policy (`IOC` or `FOK`).
- Quotes violating constraints are rejected with clear machine-readable reasons.
- Accepted quotes cannot pass validation if they violate requester limits.
- Policy/HTLC enforcement is tied to negotiated constraints, not only broad request max caps.
- Test coverage includes buy and sell flows.
- Test coverage includes `IOC` and `FOK` behaviors.
- Test coverage includes backward compatibility with old peers.
- Test coverage includes volatility and timing edge cases.

## Open Questions
- What is the canonical limit-price representation to avoid buy/sell interpretation errors?
- Should both min and max rate bounds be supported initially, or only one directional bound per order type?
- Do we extend `QuoteRespStatus`, add new reject codes, or both?
- Is explicit fill quantity required in the first delivery, or can it follow once request-side limits are live?
- Should passive offer filtering and limit-order checks be merged into one decision pipeline?

### PR DESCRIPTION
**Pull request by jtobin**
Depends on #2049. Partially resolves #2004 (remainder of workstream C, and essence of workstream D).

Adds fill quantity negotiation to the RFQ wire protocol, such that a responder can propose a fill amount smaller than a requester's maximum, with validation against other limit order constraints at the quote acceptance level (including FOK execution semantics, which was added, but not meaningfully enforced, in #2049). HTLC forwarding policies are capped at the (persisted) negotiated amount so that the channel enforces the agreed-upon limit as a max bound.

There's also some substantial refactoring done around the limit order constraint logic via the RequestConstraints interface.

Here's an illustrative use of the feature from an e2e regtest demo:

```
Frank requests up to 1,000 USDX, min 200. Alice's pilot caps the fill at
500 — above the minimum, so the partial fill is accepted under IOC
semantics.

Fill cap:    500 USDX (pilot constraint)
Requested:   min=200, max=1000 USDX
Market rate: $66,805.00/BTC
✓ Accepted
  RFQ ID: VmojZCspNmgW...
  SCID:   17831787277827515184
  Rate:   $66,805.00/BTC
  Amount: 1000 USDX
  Fill:   min=200, max=1000 USDX
  Cap:    500 USDX (pilot fillcap)

→ AddInvoice (rfq_id: VmojZCspNmgW...)
  500 USDX from Frank, reusing negotiated quote
  Quote confirmed: VmojZCspNmgW...
  Invoice: lnbcrt74844690p1p5ulzcqpp5qnsy5r868n9ql4...

→ Payment: Alice → Eve → Frank
  ✓ Payment succeeded
```

### CODE REVIEW

File: docs/release-notes/release-notes-0.8.0.md [hunk 1]
```diff
@@ -127,9 +127,18 @@
   (Immediate-Or-Cancel, the default) allows partial fills above the
   minimum, while FOK (Fill-Or-Kill) requires the full requested amount
   or rejects the quote. FOK viability is checked in `VerifyAcceptQuote`
-  with a new `FOK_NOT_VIABLE` reject code. New fields are optional and
+  with a new `FOK_NOT_VIABLE` reject code, and is enforced
```
Comment (darioAnongba) [inline docs/release-notes/release-notes-0.8.0.md:130]: Too much verbosity on these release notes IMO. 90% of the details are not meant for users or too complex. I'd remove everything and just add a single entry explaining the new feature clearly and how to use it. The fact it was split in multiple PRs is irrelevant for the users.
We should be able to add a directive to the CLAUDE.MD so agents are more concise.

File: itest/custom_channels/limit_constraints_test.go [hunk 2]
```diff
@@ -260,7 +261,7 @@ func testCustomChannelsLimitConstraints(_ context.Context,
 					AssetId: assetID,
 				},
 			},
-			PaymentMaxAmt: 180_000_000,
+			PaymentMaxAmt: invoiceMsat,
```
Comment (darioAnongba) [inline itest/custom_channels/limit_constraints_test.go:264]: Already in discussion but this shows an MPP regression. The floor is enforced per-HTLC, but the ceiling is cumulative. Only a single HTLC of exactly `max` works, two HTLCs of `max/2` each fail the floor.

Not sure what's the conclusion of the discussion but if we want FOK to be non-MPP, then we can enforce this explicitly?

What about moving the floor to `CurrentAmountMsat + htlc.AmountOutMsat` with a separate "payment complete" terminator so MPP is not broken?

cc: @GeorgeTsagk

File: tapdb/sqlc/migrations/000055_rfq_accepted_max.down.sql [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    chan_id_in BIGINT NOT NULL,
+    chan_id_out BIGINT NOT NULL,
+    htlc_id BIGINT NOT NULL,
+    asset_amt BIGINT NOT NULL,
+    amt_in_msat BIGINT NOT NULL,
+    amt_out_msat BIGINT NOT NULL,
+    UNIQUE(chan_id_in, htlc_id)
+);
+
+INSERT INTO forwards (
```
Comment (darioAnongba) [inline tapdb/sqlc/migrations/000055_rfq_accepted_max.down.sql:79]: Usual problem, the `INSERT` does not preserve the auto-increment id.

File: rfq/portfolio_pilot.go [hunk 4]
```diff
@@ -507,93 +539,119 @@ func isFOK(p fn.Option[rfqmsg.ExecutionPolicy]) bool {
...
+	return ValidAcceptQuoteRespStatus
+}
 
-	default:
-		log.Warnf("checkMinFill: unhandled request type %T",
-			req)
+// checkFillConstraints verifies that the negotiated fill amount (if
+// present) is compatible with the requester's min-fill and FOK
+// constraints.
+func checkFillConstraints(req rfqmsg.Request,
```
Comment (darioAnongba) [inline rfq/portfolio_pilot.go:613]: Legit issue found by auditor agent:

>  `checkFillConstraints` validates `fill >= MinAmount` and (FOK-only)
> `fill >= MaxAmount` but never `fill <= MaxAmount`. A malicious responder
> sending `AcceptedMaxAmount = 2^64-1` for a 6-unit request passes
> `VerifyAcceptQuote` with `ValidAcceptQuoteRespStatus`.
> 
> Downstream the policy constructor clamps (so HTLCs are safe), but the
> unclamped value persists in:
> - `MarshalAcceptedBuyQuote` / `...SellQuote` → RPC clients see
>   `assetMaxAmount=6, acceptedMaxAmount=1e18`;
> - `tapdb` persistence round-trips the adversarial value;
> - `InternalPortfolioPilot.VerifyAcceptQuote` surfaces a non-error status.

Looks safe to add a check as:

```go
if fillAmt > c.MaxAmount {
    return FillExceedsMaxQuoteRespStatus
}
```
And then forward that error on the layers and map it to a new RPC enum.

**Review by darioAnongba — COMMENTED** [review]
Pending HTLC discussion for MPP on FOK path.

File: docs/release-notes/release-notes-0.8.0.md (hunk 1 above)
Comment (jtobin) [reply docs/release-notes/release-notes-0.8.0.md:130]: Agreed, will edit.

File: itest/custom_channels/limit_constraints_test.go (hunk 2 above)
Comment (jtobin) [reply itest/custom_channels/limit_constraints_test.go:264]: Yeah after looking at this again, I think you're right that we don't actually have to sacrifice MPP. Let me take another pass over it..

File: tapdb/sqlc/migrations/000055_rfq_accepted_max.down.sql (hunk 3 above)
Comment (jtobin) [reply tapdb/sqlc/migrations/000055_rfq_accepted_max.down.sql:79]: Nice one. Will fix.

File: itest/custom_channels/limit_constraints_test.go (hunk 2 above)
Comment (jtobin) [reply itest/custom_channels/limit_constraints_test.go]: Right, so I think it's not necessary to floor the HTLC amount after all. I think the whole detour was a bit of a red herring.

The payment is ultimately going to be requested by way of an invoice anyway, and that will be subject to the standard MPP stuff at the HTLC level, namely: don't cumulatively exceed the requested amount (to be pedantic, the policy amount), and lnd will fail the whole thing atomically if we don't get it all. The whole thing, under an invoicing scheme, is basically FOK by design.

There's actually no need to have differing HTLC enforcement than in the IOC (immediate-or-cancel) case. IOC allows renegotiation of the fill amount by the responder, that's all. Once the fill quantity is decided then you still have to get it all, or fail the payment.

I'll drop those couple of commits that added the HTLC floor.

File: docs/release-notes/release-notes-0.8.0.md (hunk 1 above)
Comment (jtobin) [reply docs/release-notes/release-notes-0.8.0.md]: Resolved in latest changeset. I plan to take a clean-up pass over the release notes prior to release in order to prune them down significantly for human consumption, more generally.

File: rfq/portfolio_pilot.go (hunk 4 above)
Comment (jtobin) [reply rfq/portfolio_pilot.go:613]: Fixed in latest changeset.

File: itest/custom_channels/limit_constraints_test.go (hunk 2 above)
Comment (jtobin) [reply itest/custom_channels/limit_constraints_test.go]: Resolved in the latest changeset.

**Review by darioAnongba — APPROVED** [review]

File: rfq/portfolio_pilot_rpc.go [hunk 5]
```diff
@@ -165,7 +165,12 @@ func (r *RpcPortfolioPilot) ResolveRequest(ctx context.Context,
 				err)
 		}
 
-		return NewAcceptResolveResp(*assetRate), nil
+		fillAmount := fn.None[uint64]()
+		if resp.AcceptedMaxAmount > 0 {
+			fillAmount = fn.Some(resp.AcceptedMaxAmount)
+		}
+
+		return NewAcceptResolveResp(*assetRate, fillAmount), nil
```
Comment (GeorgeTsagk) [inline rfq/portfolio_pilot_rpc.go:173]: don't we need to `checkAllConstraints` here too?

Comment (GeorgeTsagk) [reply rfq/portfolio_pilot_rpc.go:173]: should also add test coverage for it

**Review by GeorgeTsagk — APPROVED** [review]
Lgtm

only have 1 comment for a missing check, not sure if intentional or not

File: rfq/portfolio_pilot_rpc.go (hunk 5 above)
Comment (jtobin) [reply rfq/portfolio_pilot_rpc.go:173]: That's an interesting one. After thinking about it for a few minutes, I would say: probably not, at least at the RPC level. The responder doesn't strictly need to check constraints, and it's not really their duty to do so. The onus is morally on the requester to confirm that the quote matches the constraints they set. *Caveat emptor!*

At present the internal (i.e. default) portfolio pilot _does_ call checkAllConstraints in its implementation of ResolveRequest. This is like a "courtesy optimization" to the requestor, in that it saves them the hassle of having to reject a quote that doesn't meet the constraints they set. But it's an implementation detail: a pilot might also want to return a quote that intentionally doesn't meet the constraints in a "take it or leave it" fashion, and one could argue that's courteous in a certain sense as well.

So yeah, I would say we probably _don't_ want to enforce it at the RPC level, and just leave it to the pilot implementor instead.
